app/services/bulk_punches_service.rb

```diff
@@ -0,0 +1,22 @@
+class BulkPunchesService
```

Comment:
The concept to use a service object is just to take a single action, and process the data.
But for your case, you need to validate your data too.
Prefer to use the Service Object pattern when you are ensured about your data, and avoid it when you're passing params data from a user request.

Take a look at the Form Object pattern. An inspiration: https://jaryl.medium.com/disciplined-rails-form-object-techniques-patterns-part-2-12b8d530143d

Reply:
Thanks, I didn't know about this pattern, so I implemented it.